aiida_cp2k/utils.py

```diff
@@ -74,23 +89,47 @@ def _render_section(output, params, indent=0):
                   &END KIND
         """
 
+        if six.PY2:
+            from collections import Mapping, Sequence
+        else:
+            from collections.abc import Mapping, Sequence
+
         for key, val in sorted(params.items()):
+            # the `_` is reserved for section params and evaluated in the prior call
+            if key == "_":
+                continue
+
+            # keys are not case-insensitive, ensure that they follow the current scheme
             if key.upper() != key:
-                raise ValueError("keyword '%s' not upper case" % key)
-            if key.startswith("@") or key.startswith("$"):
-                raise ValueError("CP2K preprocessor not supported")
-            if isinstance(val, dict):
-                line = "%s&%s" % (" " * indent, key)
+                raise ValueError("keyword '{key}' not upper case".format(key=key))
+
+            if key.startswith(("@", "$")):
+                raise ValueError("CP2K preprocessor directives not supported")
+
+            ispace = " " * indent
+
+            if isinstance(val, Mapping):
+                line = "{ispace}&{key}".format(ispace=ispace, key=key)
                 if "_" in val:  # if there is a section parameter, add it
-                    line += " %s" % val["_"]
+                    line += " {}".format(val["_"])
```

Comment:
The problem with a `val.pop(...)` inside a nested function call is that it is hard to grasp that adding an entry to a list changes some other datastructure. In fact, I lost quiet some time to figure out at which point the function was actually modifying the dictionary itself.

Hence, one rule in programming is that you're usually trying to avoid side-effects in function calls (the function call here being the `output.append(...)`, the side-effect that inside you're modifying a deeply nested dictionary which you got by reference).